File: packages/nats-client-py/src/klodi_nats_client/backoff.py
"""Exponential reconnect backoff with jitter — P2-5.

Mirrors ``packages/nats-client-ts/src/backoff.ts`` and
``packages/nats-client-rs/src/backoff.rs`` so the three NATS clients
agree on retry pacing under sustained outage. Per **R § cross-cutting
theme 2** symmetry across the clients is asserted by code-review today
— this module makes the contract explicit and unit-testable.

The Python ``nats-py`` library does not expose a per-attempt delay
callback (its reconnect loop uses a flat ``reconnect_time_wait`` value).
We therefore expose ``compute_backoff_seconds`` for callers that wrap
their initial connect in an outer retry loop, plus a ``BackoffPolicy``
mutable counter for integrators that want a stateful cursor.
"""

from __future__ import annotations

import random as _random
from dataclasses import dataclass
from typing import Callable, Final


# Defaults match the TS / Rust analogs by spec. See **R § P2-5**.

#: First-attempt delay, in seconds.
DEFAULT_BASE_SECONDS: Final[float] = 0.250

#: Multiplicative factor between attempts.
DEFAULT_MULTIPLIER: Final[float] = 2.0

#: Upper bound on the deterministic delay before jitter, in seconds.
DEFAULT_CAP_SECONDS: Final[float] = 60.0

#: Symmetric jitter as a fraction of the (capped) delay.
DEFAULT_JITTER_RATIO: Final[float] = 0.25


def compute_backoff_seconds(
    attempt: int,
    *,
    base_seconds: float = DEFAULT_BASE_SECONDS,
    multiplier: float = DEFAULT_MULTIPLIER,
    cap_seconds: float = DEFAULT_CAP_SECONDS,
    jitter_ratio: float = DEFAULT_JITTER_RATIO,
    rng: Callable[[], float] = _random.random,
) -> float:
    """Compute the delay (in seconds) for a 1-based reconnect ``attempt``.

    Examples (deterministic, no jitter):

    * ``attempt=1`` → ``0.250``
    * ``attempt=2`` → ``0.500``
    * ``attempt=3`` → ``1.000``
    * ``attempt=9`` → ``64.000`` → capped to ``60.000``

    Jitter samples uniformly from ``[-jitter_ratio, +jitter_ratio]`` of
    the capped value, so the post-jitter delay can briefly exceed the
    cap by up to ``jitter_ratio * cap_seconds``. The cap is a target
    ceiling for the deterministic component, not a hard cap on any
    single delay.

    :raises ValueError: when ``attempt < 1``.
    """
    if attempt < 1:
        raise ValueError(
            f"compute_backoff_seconds: attempt must be >= 1, got {attempt}"
        )
    raw = base_seconds * (multiplier ** (attempt - 1))
    capped = min(raw, cap_seconds)
    # Uniform sample in [-jitter_ratio, +jitter_ratio].
    jitter_fraction = (rng() * 2.0 - 1.0) * jitter_ratio
    delay = capped * (1.0 + jitter_fraction)
    # Floor at 0 — negative jitter on a small base shouldn't underflow
    # into a busy-loop.
    return max(0.0, delay)


@dataclass
class BackoffPolicy:
    """Stateful cursor wrapping ``compute_backoff_seconds``.

    Use case: outer retry loop that needs to remember "what attempt are
    we on?" and reset on success without the caller threading the
    counter through. The counter advances on every ``next_delay()`` call
    and resets to 0 on ``reset()``.

    Mirrors the `KlodiClient.reconnectAttempt` cursor in the TS client
    and the analogous field in Rust's connect path.
    """

    base_seconds: float = DEFAULT_BASE_SECONDS
    multiplier: float = DEFAULT_MULTIPLIER
    cap_seconds: float = DEFAULT_CAP_SECONDS
    jitter_ratio: float = DEFAULT_JITTER_RATIO
    rng: Callable[[], float] = _random.random
    _attempt: int = 0

    def next_delay(self) -> float:
        """Advance the attempt counter and return the next delay."""
        self._attempt += 1
        return compute_backoff_seconds(
            self._attempt,
            base_seconds=self.base_seconds,
            multiplier=self.multiplier,
            cap_seconds=self.cap_seconds,
            jitter_ratio=self.jitter_ratio,
            rng=self.rng,
        )

    def reset(self) -> None:
        """Reset the attempt counter to 0 — call after a successful connect."""
        self._attempt = 0

    @property
    def attempt(self) -> int:
        """Current attempt number (0 before the first ``next_delay()``)."""
        return self._attempt


__all__ = [
    "BackoffPolicy",
    "DEFAULT_BASE_SECONDS",
    "DEFAULT_CAP_SECONDS",
    "DEFAULT_JITTER_RATIO",
    "DEFAULT_MULTIPLIER",
    "compute_backoff_seconds",
]

File: packages/nats-client-ts/src/backoff.ts
/**
 * Exponential reconnect backoff with jitter — P2-5.
 *
 * The nats-core defaults are flat (`reconnectTimeWait: 2_000`) which
 * thunders the herd when a multi-replica server recovers from a
 * multi-hour outage: every client retries on the same 2s grid. This
 * module computes per-attempt delays as
 *
 *   base * multiplier^(attempt - 1)
 *
 * capped at `capMs`, then perturbed by a uniform jitter in
 * `[-jitterRatio, +jitterRatio]` of the computed value. The cap is
 * applied **before** jitter so the post-jitter value can briefly exceed
 * the cap by up to `jitterRatio * capMs` — that's intentional, the cap
 * is a target ceiling for the deterministic component, not a hard cap on
 * any single delay.
 *
 * Mirrors the Python `klodi_nats_client.backoff` and Rust
 * `klodi_nats_client::backoff` helpers — shared semantics across all
 * three clients per the cross-language symmetry contract (R § theme 2).
 */

export interface BackoffConfig {
  /** First-attempt delay in milliseconds. Default 250ms. */
  readonly baseMs: number;
  /** Per-attempt multiplier. Default 2 (doubles each attempt). */
  readonly multiplier: number;
  /**
   * Upper bound on the deterministic delay before jitter, in
   * milliseconds. Default 60_000 (60s). The post-jitter delay can
   * briefly exceed this by up to `jitterRatio * capMs`.
   */
  readonly capMs: number;
  /**
   * Symmetric jitter as a fraction of the (capped) delay. Default 0.25.
   * The actual jitter is sampled uniformly from [-jitterRatio, +jitterRatio].
   */
  readonly jitterRatio: number;
  /** Random source. Defaults to `Math.random`. Injectable for tests. */
  readonly random?: () => number;
}

export const DEFAULT_BACKOFF_CONFIG: Readonly<Omit<BackoffConfig, "random">> = {
  baseMs: 250,
  multiplier: 2,
  capMs: 60_000,
  jitterRatio: 0.25,
};

/**
 * Compute the delay (in ms) for a given reconnect `attempt` (1-based).
 *
 * Examples (deterministic / pre-jitter):
 *   attempt=1 → 250ms
 *   attempt=2 → 500ms
 *   attempt=3 → 1000ms
 *   ...
 *   attempt=9 → 64000ms → capped to 60000ms
 *
 * Jitter adds ±25% (default) on top of the capped value.
 *
 * @throws if `attempt < 1`.
 */
export function computeBackoffMs(
  attempt: number,
  config: Partial<BackoffConfig> = {},
): number {
  if (!Number.isInteger(attempt) || attempt < 1) {
    throw new Error(
      `computeBackoffMs: attempt must be a positive integer, got ${attempt}`,
    );
  }
  const merged = { ...DEFAULT_BACKOFF_CONFIG, ...config };
  const random = config.random ?? Math.random;
  const exponent = attempt - 1;
  const raw = merged.baseMs * Math.pow(merged.multiplier, exponent);
  const capped = Math.min(raw, merged.capMs);
  // Uniform sample in [-jitterRatio, +jitterRatio].
  const jitterFraction = (random() * 2 - 1) * merged.jitterRatio;
  const delay = capped * (1 + jitterFraction);
  // Floor at 0 — negative jitter on a small base shouldn't underflow
  // into a busy-loop (the library re-checks immediately).
  return Math.max(0, delay);
}
